[Feat] document tips to tune BPMN performance (#72)
libremfg/libremfg.github.io: content/how-to/bpmn/tune-performance.md

---
title: 'Tune BPMN performance'
date: '2024-02-09T09:47:47-03:00'
categories: ["how-to"]
description: Tips debug and improve the performance of your BPMN process
weight: 300
menu:
  main:
    parent: howto-bpmn
    identifier:
    name: Tune performance
---

This page documents some tips to debug [BPMN workflows]({{< relref "/how-to/bpmn/create-workflow" >}}) and improve their performance.

Manufacturing events can generate a vast amount of data.
Rhize has no limits on the data it can ingest,
and a BPMN workflow can have any number of logical flows and data transformations.
So an inefficient BPMN process can introduce performance degradations.

When writing and debugging a BPMN workflow, consider these tips.

## Avoid parallel gateways

> Running processes in [parallel]({{< relref "/how-to/bpmn/bpmn-elements#parallel-gateway" >}}) can increase the workflow's complexity by an order of magnitude.

Author: @tomhollingworth Do you have some advice on how I can get more specific here? What would the main effects be? Over taxed CPU? Out of memory?

Reviewer: Increased memory usage of NATS

caption="<em><small>The material lot object is dominating the size of this JSON payload. This a good place to start looking for optimizations.</small></em>"
>}}

## Use the JSONata extension

JSONata transformations perform better when they are precise.
A strategy to debug and minimize necessary computation is to break transformations down into smaller steps.

If you use Visual Studio Code, consider the [`jsonata-language`](https://marketplace.visualstudio.com/items?itemName=bigbug.vscode-language-jsonata) extension.
Similar to a Jupyter notebook, the extension provides an interactive environment to write JSONata expressions and to pass the outputs from one expression into the input of another.
Besides its benefit for monitoring performance, we have used it to incrementally build complex JSONata in a way that we can  document and share (in the style of literate programming).

## Manage the process context size

{{< notice "note" >}}
The max size of the process variable context comes from the default max payload size of NATS Jetstreams.
To increase this size, change your NATS configuration.
{{< /notice >}}

By default, the size of the {{< abbr "process variable context" >}} is 1MB.

> If the sum size of all variables exceeds this limit, the BPMN process exits.

Author: ```suggestion
If the sum size of all variables exceeds this limit, the BPMN process fails to execute.
```

Be mindful of the overall size of your variables, especially when outputting to new variables.

> For example, imagine an initial JSON payload, `data`, that is 600MB.

Author: ```suggestion
For example, imagine an initial JSON payload, `data`, that is 600KB.
```

> If a JSONata task slightly modifies and outputs it to a new variable, `data2`, the process variable context will exceed 1000MB and the BPMN process will exit.

Author: ```suggestion
If a JSONata task slightly modifies and outputs it to a new variable, `data2`, the process variable context will exceed 1MB and the BPMN process will exit.
```

To work around this constraint, you can save memory by mutating variables.

> That is, instead of outputting a new variable, you can output the transformed payload to the original variable name.

Author: ```suggestion
That is, instead of outputting a new variable, you can output the transformed payload to the original variable name.

Additionally, in service tasks that call APIs, use the **Response Transform Expression** to minimize the returned data to only the necessary fields. Rhize stores only the what's returned in the expression and discards the other part of the response. This is especially useful in service tasks that [Call a REST API](https://docs.rhize.com/how-to/bpmn/bpmn-elements/#call-rest-api), since you cannot precisely specify the fields in the response (as you can with a GraphQL query).
```

Author: @Jarrah-libremfg what do you think of this?

Reviewer 2: Looks good to me.

If you use Visual Studio Code, consider the [`jsonata-language`](https://marketplace.visualstudio.com/items?itemName=bigbug.vscode-language-jsonata) extension.
Similar to a Jupyter notebook, the extension provides an interactive environment to write JSONata expressions and to pass the outputs from one expression into the input of another.
Besides its benefit for monitoring performance, we have used it to incrementally build complex JSONata in a way that we can  document and share (in the style of literate programming).

## Manage the process context size

{{< notice "note" >}}
The max size of the process variable context comes from the default max payload size of NATS Jetstreams.

> To increase this size, change your NATS configuration.

Reviewer 2: Might be worth linking to the NATS docs here. They have some discussion about reasonable max sizes. The key part is while it'll go up to 64MB (IIRC) they don't recommend going above 8.

Author: Definitely a good idea, but are you able to find the link?

My quick search find only these two pages mentioning `maxMsgSize`:

- https://docs.nats.io/nats-concepts/jetstream/streams#retentionpolicy
- https://docs.nats.io/using-nats/developer/develop_jetstream/model_deep_dive

But I couldn't find that specific recommendation.

{{< /notice >}}

By default, the size of the {{< abbr "process variable context" >}}) is 1MB.

> If the sum size of all variables exceeds this limit, the BPMN process exits.

Reviewer: BPMN process exits. -> BPMN process fails to execute.

Be mindful of the overall size of your variables, especially when outputting to new variables.

> For example, imagine an initial JSON payload, `data`, that is 600MB.

Reviewer: 600KB

> If a JSONata task slightly modifies and outputs it to a new variable, `data2`, the process variable context will exceed 1000MB and the BPMN process will exit.

Reviewer: 1MB

To work around this constraint, you can save memory by mutating variables.

> That is, instead of outputting a new variable, you can output the transformed payload to the original variable name.

Reviewer 2: You can also use the `responseTransformExpression` to minimize the data returned by large requests in GraphQL or REST nodes. Only the output of the transform will be stored, not the original response.

Author: ![image](https://github.com/libremfg/libremfg.github.io/assets/47385188/6f7ee2d8-a34b-4509-b603-6fa24d7343e5)

This right? But, just to be sure―any variables that came before will still occupy memory in the context space. It's a good tip though to keep the blast radius down.

Reviewer 2: The way it works is that the original response never goes on the variable context if there is a response transform expression, only the output of that expression.

`var` above refers to where the relevant **one** of them will go on completion of the job.
